docs/data/data-grid/tree-data/tree-data.md

```diff
@@ -115,17 +115,46 @@ You can limit the sorting to the top-level rows with the `disableChildrenSorting
 
 ## Children lazy-loading
 
-:::warning
-This feature isn't implemented yet. It's coming.
+To lazy-load tree data children, set the `rowsLoadingMode` prop to `server` and listen to the `fetchRowChildren` event or pass a handler to `onFetchRowChildren` prop. It is fired when user tries to expand a row, it recieves the parent row object and a `helpers` object as parameters which has the following signature.
 
-👍 Upvote [issue #3377](https://github.com/mui/mui-x/issues/3377) if you want to see it land faster.
-:::
+```tsx
+interface GridTreeDataLazyLoadHelpers {
+  success: (rows: GridRowModel[]) => void;
+  error: () => void;
+}
+
+interface GridFetchRowChildrenParams {
+  row: GridRowModel | undefined;
+  helpers: GridTreeDataLazyLoadHelpers;
+}
+```
+
+The `onFetchRowChildren` handler is fired when the data for a specific row is requested, use it to fetch the data and call `helpers.success(newRows)` and `helpers.error()` respectively in case of success or error to let the grid update the related internal states.
+
+To enable lazy-loading for a given row, you also need to set the `isServerSideRow` prop to a function that returns `true` for the rows that have children and `false` for the rows that don't have children. If you have the information on server, you can provide an optional `getDescendantCount` prop which returns the number of descendants for a parent row.
+
+```tsx
+async function onFetchRowChildren({ row, helpers }: GridFetchRowChildrenParams) {
+  try {
+    const childRows = await fetchRows(row);
+    helpers.success(childRows);
+  } catch (error) {
+    helpers.error();
+  }
+}
 
-Alternatively, you can achieve a similar behavior by implementing this feature outside the component as shown below.
-This implementation does not support every feature of the data grid but can be a good starting point for large datasets.
+<DataGridPro
+  {...otherProps}
+  treeData
+  getTreeDataPath={getTreeDataPath}
```

Comment:
When using client-side tree data, we use `getTreeDataPath` to tell the grid how to create the tree structure.

When fetching rows on the server, I think we don’t need `getTreeDataPath` anymore.
The grid fetches the top-level rows and creates the tree structure based on the returned rows. From `isServerSideRow` we know which rows have children, and once the parent row is expanded and the children rows are fetched - we know where to put them without `getTreeDataPath`. Does this make sense?

Reply:
I had an impression the user can provide rows with any nested level in the beginning, it doesn't necessarily have to be top-level, also, the server should have stored the path in the database to be able to reflect the requests containing request payload similar to:
```tsx
{
  path: ['level 1', 'level 2', ...],
  ...otherProperties
}
```
So it should be possible for the BE servers to return the path with each row whose key the user can define using `getTreeDataPath`. Do you see another benefit of handling this internally other than the user not having to provide an extra prop?

Reply:
Options (2) and (3) combine both client-side and server-side data fetching, right? I don't think we should mix these fetching strategies. If you have some data cached on the client already, you should be able to skip the server request in `onFetchRowChildren` and return the cached data immediately.

> Do you see another benefit of handling this internally other than the user not having to provide an extra prop?

My main point was avoiding client-side row grouping in the server-side mode because the tree structure is already known.